File: android/app/src/main/java/com/tanay/echo/transcription/Whisper.kt
package com.tanay.echo.transcription

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

// Whisper transcription client — a faithful Kotlin port of src/main/transcription/whisper.ts,
// including its retry/timeout policy. parseWhisperText + joinUrl are pure (JVM-testable); the
// OkHttp client is exercised end-to-end against MockWebServer in WhisperTest (also JVM).

class TranscriptionException(message: String, val status: Int? = null) : Exception(message)

@Serializable
private data class WhisperResponse(val text: String? = null)

private val json = Json { ignoreUnknownKeys = true }

/** Parse an OpenAI-compatible /audio/transcriptions JSON body → trimmed text. */
fun parseWhisperText(body: String): String =
    (json.decodeFromString<WhisperResponse>(body).text ?: "").trim()

fun joinUrl(base: String, path: String): String =
    base.trimEnd('/') + "/" + path.trimStart('/')

/** The Whisper `language` form value, or null to let Whisper auto-detect. Blank or "auto" ⇒ null. */
fun languageParam(raw: String): String? {
    val v = raw.trim().lowercase()
    return if (v.isEmpty() || v == "auto") null else v
}

/**
 * POST a WAV to an OpenAI-compatible `/audio/transcriptions` endpoint and return the text.
 * Retries transient failures (network errors, 5xx) so a momentary tailnet blip doesn't lose
 * a dictation; never retries a real 4xx — except that, since the bias prompt is the most
 * likely thing a server rejects, it drops the prompt and tries once more before giving up.
 * `httpClient` and `sleep` are injectable so the whole policy is unit-testable.
 */
class WhisperClient(
    private val httpClient: OkHttpClient = OkHttpClient(),
    private val sleep: suspend (Long) -> Unit = { delay(it) }
) {
    suspend fun transcribe(
        wav: ByteArray,
        baseUrl: String,
        model: String,
        apiKey: String,
        prompt: String? = null,
        language: String? = null,
        retries: Int = 2,
        timeoutMs: Long = 20_000
    ): String {
        val url = joinUrl(baseUrl, "audio/transcriptions")
        var currentPrompt = prompt?.takeIf { it.isNotEmpty() }
        var lastError: TranscriptionException? = null
        var attempt = 0
        while (attempt <= retries) {
            try {
                return postOnce(url, wav, model, apiKey, currentPrompt, language, timeoutMs)
            } catch (e: TranscriptionException) {
                lastError = e
                // 4xx is a real client error (bad key/request) — retrying won't help. But the
                // optional bias prompt may be what the server rejects, so try once more without
                // it before giving up (does not consume a retry).
                if (e.status != null && e.status in 400..499) {
                    if (currentPrompt != null) {
                        currentPrompt = null
                        continue
                    }
                    throw e
                }
                if (attempt < retries) sleep(250L * (attempt + 1))
            }
            attempt++
        }
        throw lastError ?: TranscriptionException("Whisper transcription failed")
    }

    private suspend fun postOnce(
        url: String,
        wav: ByteArray,
        model: String,
        apiKey: String,
        prompt: String?,
        language: String?,
        timeoutMs: Long
    ): String = withContext(Dispatchers.IO) {
        val form = MultipartBody.Builder().setType(MultipartBody.FORM)
            .addFormDataPart("file", "audio.wav", wav.toRequestBody("audio/wav".toMediaType()))
            .addFormDataPart("model", model)
            .addFormDataPart("response_format", "json")
            .apply { if (prompt != null) addFormDataPart("prompt", prompt) }
            .apply { if (!language.isNullOrEmpty()) addFormDataPart("language", language) }
            .build()
        val req = Request.Builder().url(url).header("Authorization", "Bearer $apiKey").post(form).build()
        val call = httpClient.newBuilder().callTimeout(timeoutMs, TimeUnit.MILLISECONDS).build().newCall(req)

        val res = try {
            call.execute()
        } catch (e: IOException) {
            throw TranscriptionException("Network error reaching Whisper: ${e.message}")
        }
        res.use {
            if (!res.isSuccessful) {
                val body = try { res.body?.string() ?: "" } catch (e: IOException) { "" }
                throw TranscriptionException("Whisper returned ${res.code}: ${body.take(200)}", res.code)
            }
            parseWhisperText(res.body?.string() ?: "")
        }
    }
}

File: src/main/transcription/whisper.ts
import type { Settings } from '@shared/types'

export interface WhisperDeps {
  fetch: typeof fetch
}

export interface WhisperOpts {
  retries?: number
  timeoutMs?: number
  delay?: (ms: number) => Promise<void>
  /** Dictionary bias prompt — nudges Whisper toward custom spellings. */
  prompt?: string
  temperature?: number
}

export class TranscriptionError extends Error {
  constructor(
    message: string,
    readonly status?: number
  ) {
    super(message)
    this.name = 'TranscriptionError'
  }
}

export function joinUrl(base: string, path: string): string {
  return base.replace(/\/+$/, '') + '/' + path.replace(/^\/+/, '')
}

/**
 * POST a WAV buffer to an OpenAI-compatible `/audio/transcriptions` endpoint and
 * return the transcribed text. Retries transient failures (network errors, 5xx) so a
 * momentary tailnet blip doesn't lose a dictation; never retries 4xx (real client
 * errors). Each attempt is bounded by a timeout.
 */
export async function transcribe(
  wav: ArrayBuffer,
  settings: Pick<Settings, 'whisperBaseUrl' | 'whisperModel'>,
  apiKey: string,
  deps: WhisperDeps = { fetch },
  opts: WhisperOpts = {}
): Promise<string> {
  const retries = opts.retries ?? 2
  const timeoutMs = opts.timeoutMs ?? 20_000
  const delay = opts.delay ?? ((ms) => new Promise((r) => setTimeout(r, ms)))
  const temperature = opts.temperature ?? 0
  const url = joinUrl(settings.whisperBaseUrl, 'audio/transcriptions')

  let prompt = opts.prompt || undefined
  let lastError: unknown
  for (let attempt = 0; attempt <= retries; attempt++) {
    try {
      return await attemptTranscribe(url, wav, settings.whisperModel, apiKey, deps, timeoutMs, prompt, temperature)
    } catch (e) {
      lastError = e
      // 4xx is a real client error (bad key, bad request) — retrying won't help.
      // Exception: the optional bias prompt may be what the server rejects, so try
      // once more without it before giving up.
      if (e instanceof TranscriptionError && e.status && e.status >= 400 && e.status < 500) {
        if (prompt) {
          prompt = undefined
          attempt--
          continue
        }
        throw e
      }
      if (attempt < retries) await delay(250 * (attempt + 1))
    }
  }
  throw lastError
}

async function attemptTranscribe(
  url: string,
  wav: ArrayBuffer,
  model: string,
  apiKey: string,
  deps: WhisperDeps,
  timeoutMs: number,
  prompt?: string,
  temperature = 0
): Promise<string> {
  const form = new FormData()
  form.append('file', new Blob([wav], { type: 'audio/wav' }), 'audio.wav')
  form.append('model', model)
  form.append('language', 'en')
  form.append('response_format', 'json')
  // Deterministic decoding — greedy, no sampling drift between identical dictations.
  form.append('temperature', String(temperature))
  if (prompt) form.append('prompt', prompt)

  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeoutMs)
  let res: Response
  try {
    res = await deps.fetch(url, {
      method: 'POST',
      headers: { Authorization: `Bearer ${apiKey}` },
      body: form,
      signal: controller.signal
    })
  } catch (e) {
    throw new TranscriptionError(`Network error reaching Whisper: ${(e as Error).message}`)
  } finally {
    clearTimeout(timer)
  }

  if (!res.ok) {
    const body = await res.text().catch(() => '')
    throw new TranscriptionError(`Whisper returned ${res.status}: ${body.slice(0, 200)}`, res.status)
  }

  const data = (await res.json()) as { text?: string }
  return (data.text ?? '').trim()
}
